libensemble/utils/runners.py

```diff
@@ -84,3 +97,117 @@ def _result(self, calc_in: npt.NDArray, persis_info: dict, libE_info: dict) -> (
+                    arr[i][key] = x[i][key]
+            return arr
+        return x
+
+    def _get_points_updates(self, batch_size: int) -> (npt.NDArray, npt.NDArray):
+        return self._to_array(self.gen.ask(batch_size)), self._to_array(self.gen.ask_updates())
+
+    def _convert_tell(self, x: npt.NDArray) -> list:
+        self.gen.tell(np_to_list_dicts(x))
+
+    def _loop_over_gen(self, tag, Work):
+        """Interact with ask/tell generator that *does not* contain a background thread"""
+        while tag not in [PERSIS_STOP, STOP_TAG]:
+            batch_size = self.batch or Work["libE_info"]["batch_size"]
+            points, updates = self._get_points_updates(batch_size)
+            if updates is not None and len(updates):  # returned "samples" and "updates". can combine if same dtype
+                H_out = np.append(points, updates)
+            else:
+                H_out = points
+            tag, Work, H_in = self.ps.send_recv(H_out)
+            self._convert_tell(H_in)
+        return H_in
+
+    def _get_initial_ask(self, libE_info) -> npt.NDArray:
+        """Get initial batch from generator based on generator type"""
+        initial_batch = self.initial_batch or libE_info["batch_size"]
+        H_out = self.gen.ask(initial_batch)
+        return H_out
+
+    def _start_generator_loop(self, tag, Work, H_in):
+        """Start the generator loop after choosing best way of giving initial results to gen"""
+        self.gen.tell(np_to_list_dicts(H_in))
+        return self._loop_over_gen(tag, Work)
+
+    def _persistent_result(self, calc_in, persis_info, libE_info):
+        """Setup comms with manager, setup gen, loop gen to completion, return gen's results"""
+        self.ps = PersistentSupport(libE_info, EVAL_GEN_TAG)
+        if hasattr(self.gen, "setup"):
+            self.gen.persis_info = persis_info  # passthrough, setup() uses the gen attributes
+            self.gen.libE_info = libE_info
+            if self.gen.thread is None:
+                self.gen.setup()  # maybe we're reusing a live gen from a previous run
+        H_out = self._get_initial_ask(libE_info)
+        tag, Work, H_in = self.ps.send_recv(H_out)  # evaluate the initial sample
+        final_H_in = self._start_generator_loop(tag, Work, H_in)
+        return self.gen.final_tell(final_H_in), FINISHED_PERSISTENT_GEN_TAG
+
+    def _result(self, calc_in: npt.NDArray, persis_info: dict, libE_info: dict) -> (npt.NDArray, dict, Optional[int]):
+        if libE_info.get("persistent"):
+            return self._persistent_result(calc_in, persis_info, libE_info)
+        return self._to_array(self.gen.ask(getattr(self.gen, "batch_size", 0) or libE_info["batch_size"]))
+
+
+class LibensembleGenRunner(AskTellGenRunner):
+    def _get_initial_ask(self, libE_info) -> npt.NDArray:
+        """Get initial batch from generator based on generator type"""
+        H_out = self.gen.ask_numpy(self.initial_batch or libE_info["batch_size"])
```

Comment:
How are these batch sizes determined. How do we support fixed and/or dynamic sizes (e.g., producing as many points as has been returned)?

Reply:
2. I thought we said you call ask_numpy if exists, else call ask() - or is that dealt with by inheriting from Generator instead of LibensembleGenerator. Is there a test with a Generator in the standardised format using list of dictionary interface? We could add one in gen_classes sampling.